File: samples/FsBulletML.Sample.Unity2D.CSharp/Assets/Scripts/BgScroll.cs
using UnityEngine;
using R3;

public class BgScroll : MonoBehaviour
{
    [SerializeField]
    private float scrollSpeed1 = 0.1f;
    private Renderer rend;

    void Awake()
    {
        rend = GetComponent<Renderer>();
    }

    void Start()
    {
        Observable.EveryUpdate(destroyCancellationToken)
            .Subscribe(_ =>
            {
                var offset = rend.material.mainTextureOffset;
                offset.y -= Time.deltaTime * scrollSpeed1;
                rend.material.mainTextureOffset = offset;
            });
    }
}

File: samples/FsBulletML2.Sample.Unity2D.FSharp/FsBulletML2.Sample.Unity2D.FSharp/BgScroll.fs
namespace FsBulletML2.Sample.Unity2D.FSharp

open UnityEngine

type BgScroll () =
  inherit MonoBehaviour ()
  [<SerializeField;DefaultValue>]val mutable public scrollSpeed : float32

  /// `Renderer` は 1 回 だけ引く。毎コマ `GetComponent` する理由は無い。
  member this.Start () =
    let r = this.GetComponent<Renderer>()
    FrameTicker.Frames
    |> subscribeUntilDestroy this (fun _ ->
        let offset = r.material.mainTextureOffset
        r.material.mainTextureOffset <-
          Vector2(offset.x, offset.y - Time.deltaTime * this.scrollSpeed))
